src/validator.rs

```diff
@@ -0,0 +1,272 @@
+use crate::{
+    compilation::JSONSchema,
+    error::{error, no_error, ErrorIterator, ValidationError},
+};
+use serde_json::{Map, Value};
+use std::fmt;
+
+pub trait Validate: Send + Sync {
+    fn build_validation_error<'a>(&self, instance: &'a Value) -> ValidationError<'a> {
+        ValidationError::unexpected(instance, &self.name())
+    }
+    fn name(&self) -> String;
+
+    #[inline(always)]
```

Comment:
I am not completely sure about `always`, the `#[inline]` version gave me ~5% improvement over big/small schemas validation in this branch

Reply:
This is actually a surprise to me, I was almost certain that I did remove all the `#[inline]` attributes as adding them just because "it feels a good idea" is generally not (except if guided by benchmarking).
I'm going to remove them and re-add them as needed after benchmarking (maybe after this landed on master)